### ISSUE TITLE
mqtt keep alive timer is not correct

### ISSUE THREAD

**Issue by lepass7** [nodemcu/nodemcu-firmware#1553]
### Expected behavior

When the esp8266 loses wifi communication it should activate the m:on("offline" ... after 120 Seconds (according to the following example)
### Actual behavior

The offline event activates randomly and after 10-15 minutes
### Test code

``` Lua
-- init mqtt client with keepalive timer 120sec
m = mqtt.Client("clientid", 120, "user", "password")

-- setup Last Will and Testament (optional)
-- Broker will publish a message with qos = 0, retain = 0, data = "offline" 
-- to topic "/lwt" if client don't send keepalive packet
m:lwt("/lwt", "offline", 0, 0)

m:on("connect", function(client) print ("connected") end)
m:on("offline", function(client) print ("offline") end)
```
### NodeMCU version

master
powered by Lua 5.1.4 on SDK 1.5.4.1(39cb9a32)
### Hardware

esp8266-12e

**Comment by NicolSpies** [comment]
I experience the same and I link this to the method used in #1555 to determine if a MQTT connection is present before trying to publish a message.

**Comment by pjsg** [comment]
From looking at the code, it appears that the timeout handling is missing. After the keepalive time it sends the PINGREQ -- which should the cause the server to send PINGRESP. The client is *supposed* to wait for up to keepalive time for the PINGRESP and abort the connection if nothing is received. However, I can't see the code that does this...

**Comment by pjsg** [comment]
If somebody can try this fix #1557 in their test environment, that would be great. It appeared to work with a simple test case....

**Comment by NicolSpies** [comment]
@pjsg, I have 3 units in the field connected via 3G that experience MQTT disconnects due to the connection. I will test on the units. Unfortunately can only get to them in a few days to upload new build.
Agh!!, I forgot, you now my challenge, I will only be able to test after a merge to dev enabling me to do a custom build.

**Comment by marcelstoer** [comment]
> Agh!!, I forgot, you now my challenge, I will only be able to test after a merge to dev enabling me to do a custom build.

Sounds vaguely familiar...For some reason the Docker image doesn't work for you? I'll send you an email with firmware of the PR.

**Comment by lepass7** [comment]
@pjsg I tried fix #1557 but unfortunately the keep alive timer is still not working correct.
